src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -839,6 +839,20 @@ public INamedTypeSymbol CreateTupleTypeSymbol(ImmutableArray<ITypeSymbol> elemen
 
         protected abstract INamedTypeSymbol CommonCreateTupleTypeSymbol(ImmutableArray<ITypeSymbol> elementTypes, ImmutableArray<string> elementNames);
 
+        /// <summary>
+        /// Returns a new INamedTypeSymbol with the given underlying type and (optional) element names.
+        /// </summary>
+        /// <remarks>
+        /// Since VB doesn't support tuples yet, this call will fail in a VB compilation.
+        /// Also, the underlying type needs to be tuple-compatible.
+        /// </remarks>
+        public INamedTypeSymbol CreateTupleTypeSymbol(INamedTypeSymbol underlyingType, ImmutableArray<string> elementNames = default(ImmutableArray<string>))
```

Comment:
When creating a tuple type symbol, it would appear the options should be either
- Create from _element types_ and (optional) _element names_, optionally specifying the _assembly_ from which the underlying `System.ValueTuple` types should be taken. This corresponds to the source operation of writing a tuple type (in the given assembly), i.e. `(int a, int b)`. When you do that you don't get to select the underlying type; it is automatically taken from the (given or source) assembly's unique `System.ValueTuple` type.
- Create directly from an _underlying type_, but without specifying element names, element types, or an assembly. This corresponds to the source operation of constructing `System.ValueTuple`, by writing an underlying type, i.e. `System.ValueTuple<int, int>`. When you do that you don't get to select the element names; they are always `Item1`, etc.

The API in this PR combines both, allowing both providing an underlying type and element names. That isn't something you can do in the language, and I don't think we should support it in the APIs. I propose the following two APIs:
```cs
        public INamedTypeSymbol CreateTupleTypeSymbol(
             INamedTypeSymbol underlyingType) ...
        public INamedTypeSymbol CreateTupleTypeSymbol(
             ImmutableArray<ITypeSymbol> elementTypes,
             ImmutableArray<string> elementNames = default(ImmutableArray<string>),
             IAsemblySymbol containingAssembly = null) ...
```